File: webapp/lib/effects.rb
# frozen_string_literal: true

require_relative "bait"

# Shared building blocks for the declarative effects defined in data/cards.yaml.
# Effects are plain data — what they target and how much they add live in the
# YAML, so both clients read the same definition rather than reimplementing
# behaviour in code.
module Effects
  # A selector: a card (room/boss) or hero matches when it satisfies every
  # dimension that is present. Dimensions:
  #   tag            - the card carries this tag
  #   type           - "trap" or "creature"/"monster" (by the room's type)
  #   bait           - the card has at least one icon of this bait type
  #   preferred_bait - the hero's preferred bait (used to match heroes)
  class Selector
    def initialize(raw = {})
      raw = raw.is_a?(Hash) ? raw : {}
      @tag = raw["tag"]
      @type = raw["type"]
      @bait = raw["bait"]
      @preferred_bait = raw["preferred_bait"]
    end

    def matches?(card)
      return false if @tag && !tagged?(card)
      return false if @type && !type_matches?(card)
      return false if @bait && !has_bait?(card)
      return false if @preferred_bait && !prefers?(card)

      true
    end

    private

    def tagged?(card)
      card.respond_to?(:tags) && card.tags.include?(@tag)
    end

    def has_bait?(card)
      card.respond_to?(:bait) && card.bait.count(@bait).positive?
    end

    def prefers?(card)
      card.respond_to?(:preferred_bait) && card.preferred_bait == Bait.normalize(@preferred_bait)
    end

    def type_matches?(card)
      case @type.to_s.downcase
      when "trap"                then card.respond_to?(:trap?) && card.trap?
      when "creature", "monster" then card.respond_to?(:creature?) && card.creature?
      else false
      end
    end
  end

  # A damage aura: +`flat` and/or +`per_point × points` to every card that
  # matches the selector.
  class Aura
    def initialize(raw = {})
      raw = raw.is_a?(Hash) ? raw : {}
      @match = Selector.new(raw["match"])
      @flat = Integer(raw["flat"] || 0)
      @per_point = Integer(raw["per_point"] || 0)
    end

    def bonus(target, points)
      return 0 unless @match.matches?(target)

      @flat + @per_point * points
    end
  end
end

File: android/app/src/main/java/com/dungeonboss/game/Effects.kt
package com.dungeonboss.game

import com.dungeonboss.model.Bait
import com.dungeonboss.model.Encounter
import com.dungeonboss.model.Hero

/**
 * Shared building blocks for the declarative effects defined in `data/cards/`.
 * Effects are plain data — what they target and how much they add live in the
 * YAML, so both clients read the same definition rather than reimplementing
 * behaviour in code. Mirrors `webapp/lib/effects.rb`.
 */
object Effects {

    /** Read an integer out of a raw YAML value (Number/String), defaulting to 0. */
    fun intOf(raw: Any?, default: Int = 0): Int = when (raw) {
        is Number -> raw.toInt()
        is String -> raw.trim().toIntOrNull() ?: default
        else -> default
    }

    @Suppress("UNCHECKED_CAST")
    fun mapOf(raw: Any?): Map<String, Any?> = (raw as? Map<String, Any?>) ?: emptyMap()

    fun listOfMaps(raw: Any?): List<Map<String, Any?>> =
        (raw as? List<*>)?.mapNotNull {
            @Suppress("UNCHECKED_CAST")
            it as? Map<String, Any?>
        } ?: emptyList()

    fun boolOf(raw: Any?): Boolean = when (raw) {
        is Boolean -> raw
        is String -> raw.equals("true", ignoreCase = true)
        else -> false
    }

    /**
     * A selector: a card (room/boss) or hero matches when it satisfies every
     * dimension that is present. Dimensions:
     *   tag            - the card carries this tag
     *   type           - "trap" or "creature"/"monster" (by the room's type)
     *   bait           - the card has at least one icon of this bait type
     *   preferred_bait - the hero's preferred bait (used to match heroes)
     */
    class Selector(raw: Map<String, Any?> = emptyMap()) {
        private val tag = (raw["tag"] as? String)?.trim()?.lowercase()
        private val type = (raw["type"] as? String)?.trim()?.lowercase()
        private val bait = raw["bait"]?.let { Bait.normalize(it.toString()) }
        private val preferredBait = raw["preferred_bait"]?.let { Bait.normalize(it.toString()) }

        fun matches(card: Any): Boolean {
            if (tag != null && !tagged(card)) return false
            if (type != null && !typeMatches(card)) return false
            if (bait != null && !hasBait(card)) return false
            if (preferredBait != null && !prefers(card)) return false
            return true
        }

        private fun tagged(card: Any): Boolean = when (card) {
            is Encounter -> card.tags.contains(tag)
            is Hero -> card.tags.contains(tag)
            else -> false
        }

        private fun hasBait(card: Any): Boolean =
            bait != null && card is Encounter && card.bait.count(bait) > 0

        private fun prefers(card: Any): Boolean =
            card is Hero && card.preferredBait == preferredBait

        private fun typeMatches(card: Any): Boolean = when (type) {
            "trap" -> card is Encounter && card.trap()
            "creature", "monster" -> card is Encounter && card.creature()
            else -> false
        }
    }

    /**
     * A damage aura: +`flat` and/or +`per_point × points` to every card that
     * matches the selector.
     */
    class Aura(raw: Map<String, Any?> = emptyMap()) {
        private val match = Effects.Selector(Effects.mapOf(raw["match"]))
        private val flat = Effects.intOf(raw["flat"])
        private val perPoint = Effects.intOf(raw["per_point"])

        fun bonus(target: Any, points: Int): Int =
            if (match.matches(target)) flat + perPoint * points else 0
    }
}
